lib/src/realm_class.dart

```diff
@@ -173,31 +173,41 @@ class Realm implements Finalizable {
     if (cancellationToken != null && cancellationToken.isCancelled) {
       throw cancellationToken.exception!;
     }
-    final realm = Realm(config);
-    StreamSubscription<SyncProgress>? subscription;
-    try {
-      if (config is FlexibleSyncConfiguration) {
-        final session = realm.syncSession;
-        if (onProgressCallback != null) {
-          subscription = session.getProgressStream(ProgressDirection.download, ProgressMode.forCurrentlyOutstandingWork).listen(onProgressCallback);
-        }
-        await session.waitForDownload(cancellationToken);
-        await subscription?.cancel();
-      }
-    } catch (_) {
-      await subscription?.cancel();
-      realm.close();
-      rethrow;
+
+    if (config is! FlexibleSyncConfiguration) {
+      final realm = Realm(config);
+      return await CancellableFuture.value(realm, cancellationToken);
     }
-    return await CancellableFuture.value(realm, cancellationToken);
+
+    _ensureDirectory(config);
+
+    final asyncOpenHandle = realmCore.createRealmAsyncOpenTask(config);
+
+    return CancellableFuture.from<Realm>(() async {
+      StreamSubscription<SyncProgress>? progressSubscription;
+      if (onProgressCallback != null) {
+        final progressController = RealmAsyncOpenProgressNotificationsController._(asyncOpenHandle);
+        final progressStream = progressController.createStream();
+        progressSubscription = progressStream.listen(onProgressCallback);
+      }
+
+      final realmHandle = await realmCore.openRealmAsync(asyncOpenHandle);
```

Comment:
Can you check test "Realm.open (flexibleSync) - cancel right after open". We cancel the operation before the completion of the `CancellableFuture`. In this case we can skip calling `realmCore.openRealmAsync(asyncOpenHandle)` even though it will return immediately.

Reply:
What do you mean "can you check?" Can you clarify what is your suggestion?

Reply:
My suggestion is to prevent calling `realmCore.openRealmAsync(asyncOpenHandle)` in case the operation has been canceled. This test is simulating exactlly canceling the operation while the `CancellableFuture` is in execution.

Reply:
fixed